hpe3parclient/client.py

```diff
@@ -4872,3 +4792,252 @@ def getScheduleStatus(self, schedule_name):
+            - INV_OPERATION_VV_TASK_CANCEL_IN_PROGRESS - Invalid operation:
+            A task involving the volume is being canceled..
+        :raises: :class:`~hpe3parclient.exceptions.HTTPConflict`
+            - INV_OPERATION_VV_PROMOTE_IN_PROGRESS - Invalid operation: Volume
+            promotion is in progress.
+        :raises: :class:`~hpe3parclient.exceptions.HTTPConflict`
+            - INV_OPERATION_VV_TUNE_IN_PROGRESS - Invalid operation: Volume
+            tuning is in progress.
+        :raises: :class:`~hpe3parclient.exceptions.HTTPBadRequest`
+            - NO_SPACE - Not Enough space is available
+        :raises: :class:`~hpe3parclient.exceptions.HTTPForbidden`
+            - NODE_DOWN - The node is down.
+        :raises: :class:`~hpe3parclient.exceptions.HTTPNotFound`
+            - NON_EXISTENT_CPG - The CPG does not exists.
+        :raises: :class:`~hpe3parclient.exceptions.HTTPNotFound`
+            - NON_EXISTENT_VOL - volume doesn't exist
+        :raises: :class:`~hpe3parclient.exceptions.HTTPForbidden`
+            - VV_IN_INCONSISTENT_STATE - The volume has an internal consistency
+            error.
+        :raises: :class:`~hpe3parclient.exceptions.HTTPForbidden`
+            - VV_IS_BEING_REMOVED - The volume is being removed.
+        :raises: :class:`~hpe3parclient.exceptions.HTTPForbidden`
+            - VV_NEEDS_TO_BE_CHECKED - The volume needs to be checked.
+        :raises: :class:`~hpe3parclient.exceptions.HTTPForbidden`
+            - VV_NOT_STARTED - Volume is not started.
+        :raises: :class:`~hpe3parclient.exceptions.HTTPForbidden`
+            - INV_INPUT_VV_IS_FPVV - A fully provisioned volume cannot be
+            compressed.
+
+        """
+        info = {'action': self.TUNE_VOLUME, 'tuneOperation': tune_operation}
+        if optional is not None and not self.compression_supported:
+            if 'compression' in optional.keys():
+                del optional['compression']
+        if optional:
+            if self.primera_supported:
+                if optional.get('conversionOperation') == self.TDVV \
+                        and optional.get('compression') is True:
+                    optional['conversionOperation'] = self.CONVERT_TO_DECO
+
+                if (optional.get('conversionOperation') == self.TDVV and
+                        optional.get('compression') is False) or\
+                   (optional.get('conversionOperation') == self.TPVV and
+                        optional.get('compression') is True):
+                    raise exceptions.HTTPBadRequest("invalid input: For\
+ compression and deduplicated volume 'conversionOperation' should be\
+ 3(TDVV) and 'compression' must be specified as true")
+                if optional.get('conversionOperation') == self.FPVV:
+                    raise exceptions.HTTPBadRequest("invalid input:\
+ 'conversionOperation' value 2(FPVV) is not supported")
+
+                if 'compression' in optional:
```

Comment:
@ansarars 
What is the behavior of primera when compression parameter is passed?
should we pop the compression and proceed for volume tuning in following case?
Consider volume on primera is thin provisioning and user has passed compression as true.
As per the flow we are poping compression and proceed with tune operation and we may end up with The "volume is already thinly provisioned" exception 
We should verify this use case and change the code if needed with proper exception message.

Reply:
> @ansarars
> What is the behavior of primera when compression parameter is passed?
> should we pop the compression and proceed for volume tuning in following case?
> Consider volume on primera is thin provisioning and user has passed compression as true.
> As per the flow we are poping compression and proceed with tune operation and we may end up with The "volume is already thinly provisioned" exception
> We should verify this use case and change the code if needed with proper exception message.

@nilangekarss please review below logic if it suffice the requirement then we can proceed with it:-
        info = {'action': self.TUNE_VOLUME, 'tuneOperation': tune_operation}
        if optional:
            if self.primera_supported:
                if optional.get('compression') is True:
                    if optional.get('conversionOperation') == self.TDVV:
                        optional['conversionOperation'] = self.CONVERT_TO_DECO
                        optional.pop('compression')
                    else:
                        # Exception invalid input: For
                        #  compression and deduplicated volume 'conversionOperation' should be
                        #  3(TDVV) and 'compression' must be specified as true
                else:
                    if optional.get('conversionOperation') == self.TDVV:
                        # Exception  invalid input: For
                        #  compression and deduplicated volume 'conversionOperation' should be
                        #  3(TDVV) and 'compression' must be specified as true
                    elif optional.get('conversionOperation') == self.TPVV:
                        # Thin volume request without compression
                        optional.pop('compression')
                    else:
                        # Exception : invalid input: 'conversionOperation' value 2(FPVV) is not supported
            info = self._mergeDict(info, optional)
        response, body = self.http.put(
            '/volumes/%s' % volName, body=info)
        return body